asmcomp/deadcode.ml

```diff
@@ -72,65 +68,40 @@ let rec deadcode i =
       { i = {i with desc = Iswitch(index, cases'); next = s.i};
         regs = Reg.add_set_array i.live arg;
         exits = Array.fold_left
-                  (fun acc c -> IntSet.union acc c.exits) s.exits dc;
+                  (fun acc c -> Int.Set.union acc c.exits) s.exits dc;
       }
   | Icatch(rec_flag, handlers, body) ->
     let body' = deadcode body in
     let s = deadcode i.next in
-    let (handlers', h_exits) =
-      List.split
-        (List.map (fun (nfail, handler) ->
-           let handler' = deadcode handler in
-           ((nfail, handler'.i), handler'.exits))
-           handlers)
+    let handlers' = Int.Map.map deadcode (Int.Map.of_list handlers) in
+    let rec add_live exit (live_exits, used_handlers) =
+      if Int.Set.mem exit live_exits then
+        (live_exits, used_handlers)
+      else
+        let live_exits = Int.Set.add exit live_exits in
+        match Int.Map.find_opt exit handlers' with
+        | None -> (live_exits, used_handlers)
+        | Some handler ->
+            Int.Set.fold add_live handler.exits
```

Comment:
This needs to be conditional on rec_flag, as before, right?

Reply:
If we assume that the input was valid, then we don't, because only correctly scoped handlers will appear.

Reply:
But we could add an assertion.

Reply:
I'll an assertion here.